File: java/code/29.java
class Solution {

    public int divide(int dividend, int divisor) {
        if(dividend == Integer.MIN_VALUE && divisor == -1) return Integer.MAX_VALUE;
        if(dividend == Integer.MIN_VALUE && divisor == 1) return Integer.MIN_VALUE;
        if(dividend == Integer.MIN_VALUE && divisor == Integer.MIN_VALUE) return 1;
        if(dividend == 0 || divisor == Integer.MIN_VALUE) return 0;
        boolean sign = (dividend > 0 && divisor > 0) || (dividend < 0 && divisor < 0);
        int ans = 0;
        if(dividend < 0) dividend = -dividend;
        if(divisor < 0) divisor = -divisor;
        int num = divisor;
        int cnt = 1;
        while(dividend - divisor >= 0){
            if (dividend - num > 0){
                dividend -= num;
                ans += cnt;
                num = num + num;
                cnt = cnt + cnt;
            }
            else{
                num = divisor;
                cnt = 1;
                ans += 1;
                dividend -= num;
            }
        }
        return (sign ? ans : -ans);
    }
}

File: C++/code/29.cpp

class Solution {
public:
    int divide(int dividend, int divisor) {
        if (dividend == INT_MIN && divisor == -1) return INT_MAX;
        int ans = 0;
        int sign = (dividend < 0 && divisor > 0) || (dividend > 0 && divisor < 0) ? -1 : 1;
        if (divisor > 0) divisor = -divisor;
        if (dividend > 0) dividend = -dividend;
        while (dividend <= divisor) {
            int cnt = -1;
            int subtractor = divisor;
            while (subtractor > INT_MIN / 2 && subtractor + subtractor >= dividend) {
                cnt += cnt;
                subtractor += subtractor;
            }
            ans += cnt;
            dividend -= subtractor;
        }
        return -sign * ans;
    }
};
